server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java

```diff
@@ -121,28 +207,189 @@ public ClusterState closeIndices(ClusterState currentState, final Index[] indice
+        private final ActionListener<Map<Index, AcknowledgedResponse>> listener;
+
+        private WaitForClosedBlocksApplied(final Set<Index> blockedIndices,
+                                           final @Nullable TimeValue timeout,
+                                           final ActionListener<Map<Index, AcknowledgedResponse>> listener) {
+            if (blockedIndices == null || blockedIndices.isEmpty()) {
+                throw new IllegalArgumentException("Cannot wait for closed block to be applied to null or empty list of blocked indices");
+            }
+            this.blockedIndices = blockedIndices;
+            this.listener = listener;
+            this.timeout = timeout;
+        }
+
+        @Override
+        public void onFailure(final Exception e) {
+            listener.onFailure(e);
+        }
+
+        @Override
+        protected void doRun() throws Exception {
+            final Map<Index, AcknowledgedResponse> results = ConcurrentCollections.newConcurrentMap();
+            final CountDown countDown = new CountDown(blockedIndices.size());
+            final ClusterState state = clusterService.state();
+            for (Index blockedIndex : blockedIndices) {
+                waitForShardsReadyForClosing(blockedIndex, state, timeout, new ActionListener<AcknowledgedResponse>() {
+                    @Override
+                    public void onResponse(final AcknowledgedResponse result) {
+                        results.put(blockedIndex, result);
+                        processIfFinished();
+                    }
+
+                    @Override
+                    public void onFailure(final Exception e) {
+                        results.put(blockedIndex, new AcknowledgedResponse(false));
+                        processIfFinished();
+                    }
+
+                    private void processIfFinished() {
+                        if (countDown.countDown()) {
+                            listener.onResponse(unmodifiableMap(results));
+                        }
+                    }
+                });
+            }
+        }
+
+        private void waitForShardsReadyForClosing(final Index index, final ClusterState state, @Nullable final TimeValue timeout,
+                                                  final ActionListener<AcknowledgedResponse> listener) {
+            final IndexMetaData indexMetaData = state.metaData().index(index);
+            if (indexMetaData == null || indexMetaData.getState() == IndexMetaData.State.CLOSE) {
+                logger.debug("index {} has been blocked before closing and is already closed, ignoring", index);
+                listener.onResponse(new AcknowledgedResponse(true));
+                return;
+            }
+            final IndexRoutingTable indexRoutingTable = state.routingTable().index(index);
+            if (indexRoutingTable == null) {
+                logger.debug("index {} has been blocked before closing but is now deleted, ignoring", index);
```

Comment:
missing routing table while having index metadata usually indicates that the index has been closed in the "old style".